### DISCUSSION TITLE
Default construction with references

### DISCUSSION THREAD

**Discussion by KeyserDSoze** [dotnet/csharplang#5379] [Language Ideas]
I start with an example with record but it's similar to have a class or a struct.

```
internal sealed record Messages(List<Message> All);
internal sealed record QuestionTable(List<Question> All, Messages Messages);

var questionTable = new QuestionTable(new(), new(new()));
```

I would construct this record in a default manner to have the same behavior of the example without using all those "new" words. Like default(QuestionTable) but with references.
For instance
```
var questionTable = new QuestionTable;
```
or with a new keyword, for instance
```
var questionTable = defaultNew(QuestionTable);
```

I found nothing if someone came out with this same idea. I think that during the debate around the default behavior (in c# 7.1 https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-7.1/target-typed-default) maybe you talked about the chance to add this feature.

**Comment by svick** [comment]
[You can do this by adding an overloaded constructor:](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEAMACFBGALAbgFgAoEsAQygwEcBXCAZ2BgHsA7AFXICMAbCDAF4MbCAHcMARXpNWnHvwAUASiKliMNsGhtyvDAwh6IAEwxQIAYxZQzAWUYNyAc0aKAMjCYAeBwyeuAHwYAIK8vKokmtpQuvqGxmYW1rZSMszsXHwQHl7A3tKMGWzBYbwANBh+AYxVji6MyiQA3iQY7RjROnppRXJZSsoYIBjAABZeiqJiKpXTU+IqykPNAL4k6+rJNvb1rmrbqYWy7PhAA===)

```c#
internal sealed record QuestionTable(List<Question> All, Messages Messages)
{
    internal QuestionTable() : this(new(), new(new())) {}
}
```

Is there any reason why this wouldn't work for you?

**Reply by KeyserDSoze** [reply]
you are right, and I know that it's possible to write in some different manners. But I don't want to solve a problem with my idea, I would write "simplier".  My idea comes out only to avoid to write something more like in your example. It's something that can happen several times, and avoid the overriding of the constructor is, for me, "more courtly". I already know that if I have two constructors I could have a problem with: "what is the right constructor to use?". In this further case I could use an attribute [defaultconstructor] for instance.


```
internal sealed record QuestionTable(List<Question> All, Messages Messages)
{
    internal QuestionTable() : this(new(), new(new())) {}
    [defaultconstructor]
    internal QuestionTable(List<Question> all) : this(all, new(new())) {}
}
```

**Reply by huoyaoyuan** [reply]
>  I already know that if I have two constructors I could have a problem with: "what is the right constructor to use?".

You should not have problems, since their signatures hint their usage. Reflection based framework can have problem, where you should add `JsonConstructor` `ImportingConstructor` for them.

**Reply by KeyserDSoze** [reply]
But if I use (in a c# world that exists a default construction with references)
```
var questionTable = new QuestionTable;
```
or
```
var questionTable = defaultNew(QuestionTable);
```
on a class/record like this one
```
internal sealed record QuestionTable(List<Question> All, Messages Messages)
{
    internal QuestionTable(List<Question> all) : this(all, new(new())) {}
}
```

The compiler may not know, nay, the compiler doesn't know which constructor has to use if there are more than one of them. How can it know if it has to use this one?
```
new QuestionTable(new(), new(new()));
```
or this one?
```
new QuestionTable(new());
```

**Reply by svick** [reply]
@KeyserDSoze To me, that sounds like an argument against introducing `defaultNew` and in favor or explicitly declaring a parameterless constructor.

**Reply by KeyserDSoze** [reply]
you could be right, but in many cases you have only one constructor, and when you have more than one you could use the first one (or the constructor with more parameters than others) or the default through a default attribute. For instance with an attribute like [defaultconstructor].
```
internal sealed record QuestionTable(List<Question> All, Messages Messages)
{
    internal QuestionTable() : this(new(), new(new())) {}
    [defaultconstructor]
    internal QuestionTable(List<Question> all) : this(all, new(new())) {}
}
```
This behavior may allow the removal of the parameterless constructor, when you have 20 classes, for instance, you can remove 20 lines of code that are created only to follow the same behavior. My idea is born for not writing the same thing.
```
internal QuestionTable() : this(new(), new(new())) {}
```
